ICS 28 CCV draft w/ init and validator set update

mpoke opened the pull request:
The partial specification of Cross-Chain Validation. 

To facilitate the review process, we split the CCV spec in multiple PRs. All these PRs will be merge into the `marius/ccv` branch. Once the CCV spec is complete (all the features are included), we can create a PR to merge it to the main branch.

Features included:
- the Initialization subprotocol, that enables the creation of a unique, order channel;
- the Validator Set Update subprotocol, that enables the provider chain to inform the consumer chain of updates in the validator set.

Features not included (to be added to other PRs):
- the staking hooks that enable the timely completion of unbonding operations (e.g., undelegations) on the provider chain (i.e., waiting for the operations to mature on the consumer chain);
- the evidence / slashing subprotocol that enables the provider chain to punish validators for misbehaviors on the consumer chain;
- the distribution subprotocol that enables the validators to receive rewards for validating on the consumer chain.

sergio-mena commented on spec/app/ics-028-cross-chain-validation/README.md:
```diff
@@ -0,0 +1,53 @@
+<!-- omit in toc -->
+# Synopsis
+
+This standard document specifies packet data structure, state machine handling logic, and encoding details for Cross-Chain Validation (CCV). Note that CCV is the specific IBC level protocol that enables *Interchain Security*, a Cosmos-specific category of *Shared Security*.
```
```suggestion
This standard document specifies packet data structure, state machine handling logic, and encoding details for Cross-Chain Validation (CCV). CCV is the specific IBC level protocol that enables *Interchain Security*, a Cosmos-specific category of *Shared Security*.
```

sergio-mena replied:
(suggestion)

sergio-mena commented on spec/app/ics-028-cross-chain-validation/README.md:
```diff
@@ -0,0 +1,53 @@
+
+This standard document specifies packet data structure, state machine handling logic, and encoding details for Cross-Chain Validation (CCV). Note that CCV is the specific IBC level protocol that enables *Interchain Security*, a Cosmos-specific category of *Shared Security*.
+
+At a high level, CCV enables a *provider chain* (e.g., the Cosmos Hub) to provide *security* to multiple *consumer chains*. This means that the validator sets on the consumer chains are chosen from the validator sets of the provider chain (for more details, see the [Security Model](./overview_and_basic_concepts.md#security-model) section).
```
```suggestion
At a high level, CCV enables a *provider chain* (e.g., the Cosmos Hub) to provide *security* to multiple *consumer chains*. This means that the validator sets on the consumer chains are chosen from the validator set of the provider chain (for more details, see the [Security Model](./overview_and_basic_concepts.md#security-model) section).
```

sergio-mena commented on spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md:
```diff
@@ -0,0 +1,119 @@
+[&uparrow; Back to Outline](#outline)
+
+We consider chains that reach consensus through a proof of stake mechanism based on the model of [weak subjectivity](https://blog.ethereum.org/2014/11/25/proof-stake-learned-love-weak-subjectivity/). 
+The next block in a blockchain is *validated* and *voted* upon by a set of pre-determined *full nodes*, also known as *validators*. 
```
Careful about terminology. _full node_ and _validator_ is not exactly the same (not all full nodes are validators, at least in Tendermint terminology)

mpoke replied:
The "pre-determined *full nodes*" that "*validate* and *vote* the next block"  are  "also known as *validators*". But I get how the phrasing may be confusing.

mpoke replied:
```suggestion
The next block in a blockchain is *validated* and *voted* upon by a set of pre-determined *full nodes*; these pre-determined full nodes also known as *validators*. 
```

sergio-mena commented on spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md:
```diff
@@ -0,0 +1,119 @@
+If a validator misbehaves (e.g., validates two different blocks at the same height), its bonded tokens can be slashed. Note that the unbonding period enables the system to punish a misbehaving validator after the misbehavior is committed. 
+For more details, take a look at the [Tendermint Specification](https://github.com/tendermint/spec/blob/master/spec/core/data_structures.md) and the [Light Client Specification](https://github.com/tendermint/spec/blob/master/spec/light-client/verification/verification_002_draft.md#part-i---tendermint-blockchain).
+
+In the context of CCV, the validator sets of the consumer chains are chosen based on the tokens validators bonded on the provider chain, i.e., are chosen from the validator set of the provider chain. When these validators misbehave on the consumer chains, their bonded tokens on the provider chain is slashed. As a result, the security gained from the value of the bonded tokens on the provider chai ...
```
```suggestion
In the context of CCV, the validator sets of the consumer chains are chosen based on the tokens validators bonded on the provider chain, i.e., are chosen from the validator set of the provider chain. When these validators misbehave on the consumer chains, their bonded tokens on the provider chain are slashed. As a result, the security gained from the value of the bonded tokens on the provider chain is shared with the consumer chains. For more details, take a look at the [Interchain Security light paper](https://github.com/cosmos/gaia/blob/main/docs/interchain-security.md).
```

sergio-mena commented on spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md:
```diff
@@ -0,0 +1,119 @@
+  Once the consumer chain starts, the `InitGenesis()` method of the consumer CCV module is invoked and a client of the provider chain is created (for more details on `InitGenesis()`, take a look at the [Cosmos SDK documentation](https://docs.cosmos.network/master/building-modules/genesis.html)). 
+  For client creation, both a `ClientState` and a `ConsensusState` are necessary (as defined in [ICS 2](../../core/ics-002-client-semantics)); both are contained in the `GenesisState` of the consumer CCV module. 
+  This `GenesisState` is distributed to all operators that need to start a full node of the consumer chain (the mechanism of distributing the `GenesisState` is outside the scope of this specification).
+  > Note that although the mechanism of distributing the `GenesisState` is outside the scope of this specification, a possible approach would entail the creator of the proposal to spawn the new consumer chain to distributed the `GenesisState` via the gossip network. 
```
```suggestion
  > Note that although the mechanism of distributing the `GenesisState` is outside the scope of this specification, a possible approach would entail the creator of the proposal to spawn the new consumer chain to distribute the `GenesisState` via the gossip network. 
```

sergio-mena commented on spec/app/ics-028-cross-chain-validation/system_model_and_properties.md:
```diff
@@ -0,0 +1,171 @@
+> The following strategy is a practical example of how to ensure the Correct Relayer assumption holds. Let S denote the sending chain and D the destination chain; and let `drift(S,D)` be the time drift between S and D, i.e., `drift(S,D) =  S.currentTimestamp() - D.currentTimestamp()` (`drift(S,D) > 0` means that S is "ahead" of D). For every packet, S only sets `timeoutTimestamp = S.currentTimest ...
+
+The following assumption defines the guarantee on unbonding operations that CCV expects to be provided by the Staking module. 
+- **Staking Safety** Let `UO` be any unbonding operation that starts with an unboding transaction being executed and completes with the event that returns the corresponding stake. Let `vsc(UO)` be the VSC that contains the validator update corresponding to `UO`. Then, `UO` MUST NOT complete on the provider chain before the provider chain registers notifications of `vsc(UO)`'s maturity from all co ...
```
```suggestion
- **Staking Safety** Let `UO` be any unbonding operation that starts with an unbonding transaction being executed and completes with the event that returns the corresponding stake. Let `vsc(UO)` be the VSC that contains the validator update corresponding to `UO`. Then, `UO` MUST NOT complete on the provider chain before the provider chain registers notifications of `vsc(UO)`'s maturity from all consumer chains. 
```

sergio-mena commented on spec/app/ics-028-cross-chain-validation/technical_specification.md:
````diff
@@ -0,0 +1,1077 @@
+  }
+  ```
+
+The provider CCV module handles governance proposals to spawn new consumer chains. The structure of these proposals is defined by the `Proposal` interface in the [Governance module documentation](https://docs.cosmos.network/master/modules/gov/). The content of these proposals is described by the following interface (we omit typical fields such as tile and description):
````
```suggestion
The provider CCV module handles governance proposals to spawn new consumer chains. The structure of these proposals is defined by the `Proposal` interface in the [Governance module documentation](https://docs.cosmos.network/master/modules/gov/). The content of these proposals is described by the following interface (we omit typical fields such as title and description):
```

sergio-mena reviewed (changes requested):
Some minor/editorial comments that complement my other comments added in the PR's conversation

mpoke commented:
@sergio-mena Thanks for the feedback. I replied to all comments inline. 

> After going through the current draft, these are my comments. Please reach out to me if something is not clear
> 
> * First of all, after reading the whole spec, my main concern is that the whole system's liveness is as robust as the weakest chain. IMO, we should, at least, define a "kill_chain" method, similar to create chain that can be triggered via governance. This would be crucial in the event a chain's logic turns out to be unsound... so that the Validators can _ever_ unbond their tokens.

This is probably a good idea. @AdityaSripal was this ever discussed? Basically, a turn-off button on the provider chain :)

> 
> Here are some more detailed comments:
> 
> * "Header" section is missing. Take a look at ICS#1 and add the missing mandatory sections.

To be added before the complete draft is merge to the main branch. 

> * Questions on "Overview and Basic Concepts"
>   
>   * I understand CCV helps securing chains whose tokens don't have a big market cap. But, what would keep the motivation for CCV after those chains become successful?

The motivation of CCV is a complex, ongoing discussion. For V1, Hub Minimalism may be the best motivation. Future versions may enable higher security.  

>   * So a consumer chain MUST (?) be created by the provider chain. This means it cannot pre-exist... reason?

It can pre-exist, but not as a consumer chain. To become a consumer chain, a hard-fork is required. The operators of validators of the provider chain will start full-nodes of the consumer chain starting from a Genesis state; this state can be exported by a pre-existing chain.

>   * The terms "full node" and "validator" seem to be used interchangeably. They are not the same in Tendermint.

Fixed.

>   * Client creation. "`InitGenesis` method of the consumer CCV module". How are we sure such a method exists? MUST the consumer chain be implemented with the Cosmos SDK (alternative: App using ABCI directly)

Currently, the spec assumes that the app is implemented using Cosmos SDK. It's not really a requirement, e.g., instead of `InitGenesis`, it can be any function that is called when receiving a `InitChain` message from the consensus engine. In the Technical Specification, that's the actual precondition of `InitGenesis`. However, we should probably remove any  dependencies to Cosmos SDK.

>   * The text mentions _child chain_. Definition?

Consumer. Fixed. 

>   * So unbonding is now governed by the slowest consumer chain... what if a consumer chain gets stuck? Is this scalable?

If the consumer gets stuck, the unbonding cannot complete. Not sure I understand the second question. What do you mean by scalable?

> * Questions on "System Model and Properties"
>   
>   * "Valid Blockchain". What does this mean?: "protocols are executed correctly". Who executes them correctly? All processes? No byzantine processes allowed?

I agreee. "Valid Blockchain" is overloaded and ambiguous.  @josef-widder we should probably be more exact in what does CCV expect here, right?

>   * Correct relayer seems to not accept packet timeouts. What if one happens? Can we recover from that?

This is marked w/ a ToDo in the Technical Spec. In the case of a timeout, probably all the unbonding operations will complete. However, the spec is written under the assumption that timeouts cannot happen. There is a discussion on this under the definition of the "Correct Relayer" assumption.  

>   * "Desired properties" talks about tokens, but doesn't specify from which chain they are. I reckon it's always provider chain tokens, but it would help to clarify it in the text

Actually, it is specified, e.g., "tokens bonded by validator `val` on the provider chain".  

>   * "Unbonding Invariant" doesn't declare "cc". Shouldn't it also be part of the forall?

Actually, it does. It doesn't declare "val". It's fixed in the next iteration (https://github.com/cosmos/ibc/pull/646)

>   * "Validator Update To VSC Validity" doesn't mention slashing, should it?

No. Slashing is a feature not yet specified. But why should Validator Update To VSC Validity mention slashing?

>   * "Validator Update To VSC Liveness" you could actually relax it to "_a_ consumer chain" without loss of generality

Not really. I want it to include also Provide VSC uniformity, i.e., If the provider chain provides a VSC to a consumer chain, then it MUST eventually provide that VSC to all consumer chains.

>   * "Register Maturity" properties. No mention of aggregated VSCs. So it's not clear how these properties apply to them. An aggregated VSC may not _exactly_ match _any_ sent VSC

The consumer chain MAY aggregate multiple VSCs before applying them to the validator set. However, ACKs are sent back for every VSCPacket (i.e., for every VSC). The  "Register Maturity" properties are concern only with providing VSCs by the consumer chain and registering VSC maturity notifications on the provider chain. They don't have anything to do with aggregated VSCs.  

> * Questions on "Technical Specification"
>   
>   * Interaction with staking module. Don't we need to modify it? ... so that it doesn't enforce the local unbonding period?

The completion of unbonding operations is specified in the next iteration (https://github.com/cosmos/ibc/pull/646)

>   * Provider CCV module initialized at genesis. So how do we add support to existing chains?

Hard forks.

josef-widder commented:
"Valid Blockchain" means basically that the security model holds. In the case of Tendermint: <1/3 faulty validators. The properties, we actually need
- no forks. This ensures that a chain A header seen by B actually corresponds to a block of A
- valid transitions on A, that is, chain A executed the state machine properly. This ensures that a packet appears to be sent on chain A only if `sendPacket` has been called.

Together this ensures that 
- if a packet from A is received by B it actually has been sent by A. 
- If chain B received a validator set change, this change actually was correctly computed on A
- if an ACK packet was received on chain A, the ACK was correctly computed on B, that is, the unbonding period passed

From another viewpoint, if the provider chain does arbitrary state transitions (or if there is a fork that is installed on the consumer's light client), we cannot guarantee a lot on the consumer chain.

mpoke commented:
>"Valid Blockchain" means basically that the security model holds. In the case of Tendermint: <1/3 faulty validators. The properties, we actually need
> - no forks. This ensures that a chain A header seen by B actually corresponds to a block of A
> - valid transitions on A, that is, chain A executed the state machine properly. 

I would also add progress here, i.e., we assume both chains make progress. 

In general, the "Valid Blockchain" assumption entails that 
- the underlying consensus engine satisfies both safety and liveness (i.e., no forks and eventually add new blocks); 
- the execution of the state machine follows the described protocol (e.g., when receiving the first VSCPacket, the consumer chain sets the CCV channel to VALIDATING).

@josef-widder do you see any other requirements that are included into the  "Valid Blockchain" assumption? I agree that this assumption is necessary to ensure all the properties of IBC packets.

I do though abuse a bit the "Valid Blockchain" assumption in the argument of correctness. Probably I should point directly to the functions in the Technical Specification that describe the functionality provided by CCV.

josef-widder commented:
> @josef-widder do you see any other requirements that are included into the "Valid Blockchain" assumption? I agree that this assumption is necessary to ensure all the properties of IBC packets.

I think that is basically it. I agree that liveness also should be in the picture. But perhaps it makes sense to separate safety and liveness? In my mind "valid" was only concerned with the safety part. (As the "validity" property in typical consensus definitions). 

> I do though abuse a bit the "Valid Blockchain" assumption in the argument of correctness. Probably I should point directly to the functions in the Technical Specification that describe the functionality provided by CCV.

At least it would be good to understand which property depend on which assumptions. If this is clear, we don't need to go into too much detail in the correctness arguments.

sergio-mena commented:
> @sergio-mena Thanks for the feedback. I replied to all comments inline.

@mpoke My pleasure. I just included here comments where I have something to reply.

> > * So unbonding is now governed by the slowest consumer chain... what if a consumer chain gets stuck? Is this scalable?
> 
> If the consumer gets stuck, the unbonding cannot complete. Not sure I understand the second question. What do you mean by scalable?

Forget about the word "scalable" here. It was a poor choice on my side (I was meaning, scalable as the number of consumer chains grows). Basically, my question here is already fully addressed here and in the general comment.

> > * "Validator Update To VSC Validity" doesn't mention slashing, should it?
> 
> No. Slashing is a feature not yet specified. But why should Validator Update To VSC Validity mention slashing?

The property itself is good. My comment was rather relating to the text between parentheses "_i.e., resulted from a change in the amount of bonded tokens on the provider chain_". Maybe a way to address my comment could be to change "i.e." by "e.g." there

> > * "Validator Update To VSC Liveness" you could actually relax it to "_a_ consumer chain" without loss of generality
> 
> Not really. I want it to include also Provide VSC uniformity, i.e., If the provider chain provides a VSC to a consumer chain, then it MUST eventually provide that VSC to all consumer chains.

Fair enough. When I work with properties for specs, I try to make each of them as simple as possible, assuming their actual usefulness comes when considered together, hence my comment. But I'm lacking context here to clearly see what's best.

mpoke commented:
Added the following changes in commit https://github.com/cosmos/ibc/pull/640/commits/ab41f915115ce76b2bb3b567d1ef47ecdafbe897
- Extra assumptions needed from the staking module.
- Correctness argument of properties concerning the staking module interface.  
- Correctness argument of system invariants.

mpoke commented:
> > > * "Validator Update To VSC Validity" doesn't mention slashing, should it?
> > 
> > 
> > No. Slashing is a feature not yet specified. But why should Validator Update To VSC Validity mention slashing?
> 
> The property itself is good. My comment was rather relating to the text between parentheses "_i.e., resulted from a change in the amount of bonded tokens on the provider chain_". Maybe a way to address my comment could be to change "i.e." by "e.g." there

@sergio-mena But slashing is "a change in the amount of bonded tokens". It's a forced change, not user initiated, but a change nonetheless.

mpoke commented on spec/app/ics-028-cross-chain-validation/system_model_and_properties.md:
```diff
@@ -0,0 +1,275 @@
+The *Validator Update Provision* and *Unbonding Safety* assumptions define what is needed from the provider Staking module. 
+A more thorough discussion of the environment in which CCV operates is given in the section [Placing CCV within Cosmos SDK](./technical_specification.md#placing-ccv-within-cosmos-sdk).
+
+- ***Valid Blockchain***: Both the provider and the consumer chains are *valid*. 
+  This means that the protocols are executed correctly and the underlying consensus algorithm satisfies both safety and liveness properties. 
+  For more details, take a look at the [Tendermint Paper](https://arxiv.org/pdf/1807.04938.pdf).
```
```suggestion
- ***Safe Blockchain***: Both the provider and the consumer chains are *safe*. This means that, for every chain, the underlying consensus engine satisfies safety (e.g., the chain does not fork) and the execution of the state machine follows the described protocol. 

- ***Live Blockchain***: Both the provider and the consumer chains are *live*. This means that, for every chain, the underlying consensus engine satisfies liveness (i.e., new blocks are eventually added to the chain).

> **Note**: Both *Safe Blockchain* and *Live Blockchain* assumptions require the consensus engine's assumptions to hold, e.g., up to a third of validators can be Byzantine. For more details, take a look at the [Tendermint Paper](https://arxiv.org/pdf/1807.04938.pdf).
```

@josef-widder  do you agree with this re-def of the Valid Blockchain assumption? Clearly it must be made consistent throughout the spec.  

Note that *Live Blockchain* is needed only for liveness, while *Safe Blockchain* is needed for both, since *Correct Relayer* requires some state to be updated in order to make progress and *Correct Relayer* is needed for liveness.

sergio-mena commented:
> > > > * "Validator Update To VSC Validity" doesn't mention slashing, should it?
> > > 
> > > 
> > > No. Slashing is a feature not yet specified. But why should Validator Update To VSC Validity mention slashing?
> > 
> > 
> > The property itself is good. My comment was rather relating to the text between parentheses "_i.e., resulted from a change in the amount of bonded tokens on the provider chain_". Maybe a way to address my comment could be to change "i.e." by "e.g." there
> 
> @sergio-mena But slashing is "a change in the amount of bonded tokens". It's a forced change, not user initiated, but a change nonetheless.

That's true. Maybe less obvious when you read it, but yes the sentence is correct

AdityaSripal reviewed (commented):
Partial review. Will continue going through it in the next couple of days

mpoke commented:
The changes in https://github.com/cosmos/ibc/pull/640/commits/0d9edad883c6d2fa619b8e8bd557e2c131ccbefb remove dependencies of the CCV spec to Cosmos SDK. 

@josef-widder Do you agree that CCV is actually a module in an ABCI application that depends on functionalities provided by a Staking module (that implements PoS) and a Governance module (that allows for proposals to be voted upon)?

mpoke commented on spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md:
```diff
@@ -0,0 +1,119 @@
+
+
+
+# Security Model
```
I made some changes that I think make this section easier to read. 
- I moved the entire single chain behavior in a Background paragraph. If we add eventually a Background chapter (which we probably should), then I'll move all these Background paragraphs there. I'd not just remove them since they contain important info especially for readers that do not have a deep understanding of Cosmos. 
- I left the CCV implications on the security model (e.g., how is the unbonding period affected). I find it important to understand the rest of the specification.
- Regarding the assumptions made by CCV, these are described in the [Assumption section](https://github.com/cosmos/ibc/blob/marius/ccv-init-and-vsc/spec/app/ics-028-cross-chain-validation/system_model_and_properties.md#assumptions).

mpoke commented on spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md:
```diff
@@ -0,0 +1,119 @@
+> 
+> For more details on the planned releases, take a look at the [Interchain Security light paper](https://github.com/cosmos/gaia/blob/main/docs/interchain-security.md#the-interchain-security-stack).
+
+# Definitions and Overview
```
IMO, the two sections before it (i.e., Security Model and Motivation) are more high-level; their content is needed for a better understanding of CCV at a high-level, before we go into definitions and overview.

mpoke commented on spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md:
```diff
@@ -0,0 +1,119 @@
+Consumer chains are created through governance proposals. For details on how governance proposals work, take a look at the [Cosmos SDK documentation](https://docs.cosmos.network/master/modules/gov/).
+
+The channel initialization consists of four phases:
+- **Create clients**: The provider CCV module handles every passed proposal to spawn a new consumer chain. Once it receives a proposal, it creates a client of the consumer chain (as defined in [ICS 2](../../core/ics-002-client-semantics)). 
```
This is described in details in the t[echnical specification](https://github.com/cosmos/ibc/blob/marius/ccv-init-and-vsc/spec/app/ics-028-cross-chain-validation/technical_specification.md#ccv-pcf-ccprop1). Here is just a high-level overview.

mpoke commented on spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md:
```diff
@@ -0,0 +1,119 @@
+- **Connection handshake**: A relayer is responsible for initiating the connection handshake (as defined in [ICS 3](../../core/ics-003-connection-semantics)). 
+- **Channel handshake**: A relayer is responsible for initiating the channel handshake (as defined in [ICS 4](../../core/ics-004-channel-and-packet-semantics)). The channel handshake must be initiated on the consumer chain. The handshake consists of four messages that need to be received for a channel built on top of the expected clients. We omit the `ChanOpenAck` message since it is not relevant ...
+  - *OnChanOpenInit*: On receiving the *FIRST* `ChanOpenInit` message, the consumer CCV module sets the status of its end of the CCV channel to `INITIALIZING`.
+  - *OnChanOpenTry*: On receiving the *FIRST* `ChanOpenTry` message, the provider CCV module sets the status of its end of the CCV channel to `INITIALIZING`.
```
This is described in details in the technical spec, here https://github.com/cosmos/ibc/blob/marius/ccv-init-and-vsc/spec/app/ics-028-cross-chain-validation/technical_specification.md#ccv-pcf-cotry1 and here https://github.com/cosmos/ibc/blob/marius/ccv-init-and-vsc/spec/app/ics-028-cross-chain-validation/technical_specification.md#ccv-ccf-coinit1

mpoke commented on spec/app/ics-028-cross-chain-validation/overview_and_basic_concepts.md:
```diff
@@ -0,0 +1,119 @@
+  - *OnChanOpenInit*: On receiving the *FIRST* `ChanOpenInit` message, the consumer CCV module sets the status of its end of the CCV channel to `INITIALIZING`.
+  - *OnChanOpenTry*: On receiving the *FIRST* `ChanOpenTry` message, the provider CCV module sets the status of its end of the CCV channel to `INITIALIZING`.
+  - *OnChanOpenConfirm*: On receiving the *FIRST* `ChanOpenConfirm` message, the provider CCV module sets the status of its end of the CCV channel to `VALIDATING`.
+- **Channel completion**: Once the provider chain sets the status of the CCV channel to `VALIDATING`, it provides a VSC (i.e., validator set change) to the consumer chain (see [next section](#validator-set-update)). On receiving the *FIRST* `VSCPacket`, the consumer CCV module sets the status of its end of the CCV channel to `VALIDATING`. 
```
You mean on `ChanOpenAck` on the consumer chain, right? 

I kind of took  this design for granted since it was already there when I joined the project. But yeah, I guess that since on `ChanOpenAck` the consumer sets the channel to OPEN, it can also set it as VALIDATING. This would actually enable the consumer to send Distribution / Evidence packets before the first  VSC packet is received from the provider, which means rewarding or slashing the initial validator set. Plus, it avoids the corner case when the validator set never changes.

@josef-widder Do you see any issue with this change?

mpoke replied:
I talked with Josef about this and we agreed to have a chat with Jovan to figure out if we could actually drop the VALIDATING state completely. 

My suggestion would be:
- validators cannot unbond during the initialization protocol (no ACKs);
- as a result, the initial validator set on the consumer chain is trust worthy and the consumer chain can already accept TXs;
- a consumer chain is registered on the provider as soon as the governance proposal passes (the channel doesn't have to be initialized);
- for every valset update, send a VSC packet to all registered consumer chains w/ an open channel; if the channel is not open yet, queue the VSC packets to send them later (once the channel is initialized);
- for unbonding, any initiated unbonding operations must wait for ACKs from all registered channels (see PR https://github.com/cosmos/ibc/pull/646/)

@AdityaSripal Do you see any reason to keep the VALIDATING / INITIALIZING states?

AdityaSripal reviewed (commented):
Finished a first run-through of the spec. Great work!

mpoke commented on spec/app/ics-028-cross-chain-validation/system_model_and_properties.md:
```diff
@@ -0,0 +1,276 @@
+  let `vsc(uo)` be the VSC that contains `U(uo)`. 
+  Then,
+  - (*unbonding initiation*) the provider CCV module MUST be notified of `uo`'s initiation before receiving `U(uo)`;
+  - (*unbonding completion*) `uo` MUST NOT complete on the provider chain before the provider chain registers notifications of `vsc(uo)`'s maturity from all consumer chains.
```
"the provider chain also elapses its unbonding period as well" is not needed by CCV, but by the Staking module. Thus, is out of the scope of this specification.

mpoke commented on spec/app/ics-028-cross-chain-validation/technical_specification.md:
```diff
@@ -0,0 +1,1088 @@
+    // return an aggregated list of updates, i.e., 
+    // keep only the latest update per validator;
+    // the original list is not modified
+    Aggregate(): [ValidatorUpdate]
```
The function is implemented [here](https://github.com/cosmos/interchain-security/blob/93b9862c0864f91fa0234b9f0aec29018924e6a5/x/ccv/utils/utils.go#L7). Currently, it is called on receiving new validator updates, i.e., new updates overwrite updates of the same validator. However, this is an implementation detail.

In the spec, new validator updates are **_appended_** to the list, thus the order is preserve, i.e., more recent updates are added to the end.  Before applying these updates, we need to **_aggregate_** them, i.e., the requirement is that we apply only the most recent update per validator.

mpoke commented on spec/app/ics-028-cross-chain-validation/technical_specification.md:
```diff
@@ -0,0 +1,1088 @@
+ 
+## Sub-protocols
+
+> TODO: What about `EmitEvent()`? There are some events emitted throughout the code, but none of them are specified. 
```
I agree. We should remove them from the code then (e.g., https://github.com/cosmos/interchain-security/blob/main/x/ccv/child/module.go#L338).

mpoke commented on spec/app/ics-028-cross-chain-validation/technical_specification.md:
````diff
@@ -0,0 +1,1088 @@
+}
+```
+- Initiator: 
+  - The IBC module on the provider chain.
````
Indeed, that was the reason for having the IBC module as the initiator. I think the purpose of the "Initiator" is a bit unclear; it may be because of the name. In my view, there should be two different concepts that need to be specify for every function:
- The module that calls the function. This enables us to reason about the external dependencies (e.g., if `onChanOpenInit` is called by the IBC module, then any changes to the IBC module will affect the logic in this function).
- The event that triggered the function to be called. This enables us to reason about the correctness of the protocol (e.g., `onChanOpenInit` is called as a result of a relayer submitting a `MsgChanOpenInit`). 

We could name the first one "Caller", and the second one "Trigger". What do you think?

mpoke commented on spec/app/ics-028-cross-chain-validation/technical_specification.md:
```diff
@@ -0,0 +1,1088 @@
+    // - only ordered channels allowed
+    abortTransactionUnless(order == ORDERED)
+    // - require the portIdentifier to be the port ID the CCV module is bound to
+    abortTransactionUnless(portIdentifier == ProviderPortId)
```
I agree. We should do that also in the code, i.e., https://github.com/cosmos/interchain-security/blob/main/x/ccv/parent/module.go#L251

mpoke commented on spec/app/ics-028-cross-chain-validation/technical_specification.md:
```diff
@@ -0,0 +1,1088 @@
+    // - only ordered channels allowed
+    abortTransactionUnless(order == ORDERED)
+    // - require the portIdentifier to be the port ID the CCV module is bound to
+    abortTransactionUnless(portIdentifier == ConsumerPortId)
```
I agree. We should do that also in the code.

mpoke commented on spec/app/ics-028-cross-chain-validation/technical_specification.md:
```diff
@@ -0,0 +1,1088 @@
+function onChanCloseConfirm(
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier) {
+    // do nothing
```
I've sent you and Jehan a message on Slack regarding this. In a nutshell, I think `onChanCloseInit` should abort the transaction, i.e., do not allow users to close the channel. Then add the cleanup logic in a handler of a governance proposal, i.e., "remove consumer chain". Add same cleanup logic also on timeout on the provider.

mpoke replied:
I opened an issue in the code repo to keep track of this, i.e.,  https://github.com/cosmos/interchain-security/issues/35

mpoke commented on spec/app/ics-028-cross-chain-validation/technical_specification.md:
````diff
@@ -0,0 +1,1088 @@
+```typescript
+// CCF: Consumer Chain Function
+// implements the AppModule interface
+function InitGenesis(state: ConsumerGenesisState) {
````
Indeed. This is actually not trivial and we need to discuss it in more details. 

Currently, the `InitGenesis` of the staking module returns a list of validator updates, i.e., https://github.com/cosmos/cosmos-sdk/blob/master/x/staking/module.go#L158. `InitGenesis` has different logic depending on whether the genesis state was exported, i.e., https://github.com/cosmos/cosmos-sdk/blob/dd0d71272b/x/staking/genesis.go#L143.    

In the context of CCV, we need to make the following changes on the **consumer chain**:
- the staking `InitGenesis` does not return a list of validator updates.
- If the `GenesisState` was exported (i.e., it contains a list with the powers of the last validator set), the staking `InitGenesis` must provide this list to the CCV `InitGenesis`.
- The CCV `InitGenesis` must implement the logic from the`ApplyAndReturnValidatorSetUpdates` method of the staking module (see [here](https://github.com/cosmos/cosmos-sdk/blob/master/x/staking/keeper/val_state_change.go#L112)) to create a list of validator updates (i.e., a diff between the last validator set, if any, and the new validator set from the ProviderConsensusState). This  list of validator updates is returned to the consensus engine.

mpoke replied:
I opened an issue in the code repo to keep track of this, i.e., https://github.com/cosmos/interchain-security/issues/34

mpoke replied:
@AdityaSripal Should we fix this issue in this PR or open another one since this one is big enough?
